dep0/lib/03_typecheck/src/is_impossible.cpp

```diff
@@ -0,0 +1,169 @@
+static bool is_impossible(expr_t::i32_t const&) { return false; }
+static bool is_impossible(expr_t::i64_t const&) { return false; }
+static bool is_impossible(expr_t::u8_t const&) { return false; }
+static bool is_impossible(expr_t::u16_t const&) { return false; }
+static bool is_impossible(expr_t::u32_t const&) { return false; }
+static bool is_impossible(expr_t::u64_t const&) { return false; }
+static bool is_impossible(expr_t::boolean_constant_t const&) { return false; }
+static bool is_impossible(expr_t::numeric_constant_t const&) { return false; }
+static bool is_impossible(expr_t::string_literal_t const&) { return false; }
+static bool is_impossible(expr_t::boolean_expr_t const&);
+static bool is_impossible(expr_t::relation_expr_t const&);
+static bool is_impossible(expr_t::arith_expr_t const&);
+static bool is_impossible(expr_t::var_t const&) { return false; }
+static bool is_impossible(expr_t::global_t const&) { return false; }
+static bool is_impossible(expr_t::app_t const&);
+static bool is_impossible(expr_t::abs_t const&);
+static bool is_impossible(expr_t::pi_t const&);
+static bool is_impossible(expr_t::array_t const&) { return false; }
+static bool is_impossible(expr_t::init_list_t const&);
+static bool is_impossible(expr_t::subscript_t const&);
+static bool is_impossible(expr_t::because_t const&);
+
+static bool is_impossible(
+    std::vector<func_arg_t>::const_iterator begin,
+    std::vector<func_arg_t>::const_iterator end,
+    expr_t const& ret_type,
+    body_t const* body)
+{
+    return std::any_of(begin, end, [] (func_arg_t const& x) { return is_impossible(x.type); })
+        or is_impossible(ret_type)
+        or (body and is_impossible(*body));
+}
+
+bool is_impossible(expr_t const& x)
+{
+    return match(x.value, [] (auto const& x) { return is_impossible(x); });
+}
+
+bool is_impossible(expr_t::boolean_expr_t const& x)
+{
+    return match(
+        x.value,
+        [] (expr_t::boolean_expr_t::not_t const& x)
+        {
+            return is_impossible(x.expr.get());
+        },
+        [] (auto const& x)
+        {
+            return is_impossible(x.lhs.get()) or is_impossible(x.rhs.get());
+        });
+}
+
+bool is_impossible(expr_t::relation_expr_t const& x)
+{
+    return match(
+        x.value,
+        [] (auto const& x)
+        {
+            return is_impossible(x.lhs.get()) or is_impossible(x.rhs.get());
+        });
+}
+
+bool is_impossible(expr_t::arith_expr_t const& x)
+{
+    return match(
+        x.value,
+        [] (auto const& x)
+        {
+            return is_impossible(x.lhs.get()) or is_impossible(x.rhs.get());
+        });
+}
+
+bool is_impossible(expr_t::app_t const& x)
+{
+    return is_impossible(x.func.get())
+       or std::ranges::any_of(x.args, [] (expr_t const& x) { return is_impossible(x); });
+}
+
+bool is_impossible(expr_t::abs_t const& x)
+{
+    return is_impossible(x.args.begin(), x.args.end(), x.ret_type.get(), &x.body);
```

Comment:
This might need more thoughts. What does it mean to simply construct a lambda containing the impossible statement? Invoking it is impossible but constructing it seems fine. For example seems ok to pass it as argument to a function which later discards it.